consensus/ninja_tortoise.go

```diff
@@ -0,0 +1,456 @@
+				ni.Debug("correction vectors for g %d vector %d effective %d tally %d", g, corr.vector, corr.effectiveCount, tally)
+				tally = tally.Add(v.Multiplay(corr.effectiveCount).Add(corr.vector))
+				ni.Debug("tally for pattern %d  and block %d is %d", newMinGood.id, b, tally)
+				ni.tTally[newMinGood][b] = tally //in g's view -> in p's view
+			}
+		} else {
+			ni.Debug("no correction vectors for %", g)
+		}
+	}
+}
+
+//for all layers from pBase to i add b's votes, mark good layers
+// return new minimal good layer
+func (ni *ninjaTortoise) findMinimalGoodLayer(layer *mesh.Layer) mesh.LayerID {
+	var minGood mesh.LayerID
+
+	var j mesh.LayerID
+	if Window > layer.Index() {
+		j = ni.pBase.Layer() + 1
+	} else {
+		j = Max(ni.pBase.Layer()+1, layer.Index()-Window)
+	}
+
+	for ; j < layer.Index(); j++ {
+		// update block votes on all patterns in blocks view
+		sUpdated := ni.updateBlocksSupport(layer.Blocks(), j)
+		//todo do this as part of previous for if possible
+		//for each p that was updated and not the good layer of j check if it is the good layer
+		for p := range sUpdated {
+			//if a majority supports p (p is good)
+			//according to tal we dont have to know the exact amount, we can multiply layer size by number of layers
+			jGood, found := ni.tGood[j]
+			threshold := 0.5 * float64(mesh.LayerID(ni.LayerSize)*(layer.Index()-p.Layer()))
+
+			if (jGood != p || !found) && float64(ni.tSupport[p]) > threshold {
+				ni.tGood[p.Layer()] = p
+				//if p is the new minimal good layer
+				if p.Layer() < layer.Index() {
+					minGood = p.Layer()
+				}
+			}
+		}
+	}
+	ni.Debug("found minimal good layer %d", minGood)
+	return minGood
+}
+
+func (ni *ninjaTortoise) updateBlocksSupport(b []*mesh.Block, j mesh.LayerID) map[votingPattern]struct{} {
+	sUpdated := map[votingPattern]struct{}{}
+	for _, block := range b {
+		//check if block votes for layer j explicitly or implicitly
+		p, found := ni.tExplicit[block.ID()][j]
+		if found {
+			//explicit
+			//if _, expFound := ni.tExplicit[block.ID()]; !expFound {
+			//	ni.tExplicit[block.ID()] = make(map[mesh.LayerID]votingPattern, K*ni.LayerSize)
+			//}
+			//ni.tExplicit[block.ID()][j] = p
+			ni.tSupport[p]++         //add to supporting patterns
+			sUpdated[p] = struct{}{} //add to updated patterns
+
+			//implicit
+		} else if eff, effFound := ni.tEffective[block.ID()]; effFound {
+			p, found = ni.tPatSupport[eff][j]
+			if found {
+				ni.tSupport[p]++         //add to supporting patterns
+				sUpdated[p] = struct{}{} //add to updated patterns
+			}
+		}
+	}
+	return sUpdated
+}
+
+func (ni *ninjaTortoise) addPatternVote(p votingPattern) func(b *mesh.Block) {
+	addPatternVote := func(b *mesh.Block) {
+		var vp map[mesh.LayerID]votingPattern
+		var found bool
+		if vp, found = ni.tExplicit[b.ID()]; b.Layer() > 0 && !found {
+			panic("block has no explicit voting, something went wrong ")
+		}
+
+		for _, ex := range vp {
+			if ex.Layer() > ni.pBase.Layer() {
```

Comment:
Why do you need this if? You need to count all explicit votes in P's view even if they are for blocks that are beyond Pbase